### ISSUE TITLE
[Bug] mp_maxmoney bug

### ISSUE THREAD

**Issue by eat1k** [s1lentq/ReGameDLL_CS#357]
Adding money (I've tested with rg_add_account from ReAPI) is not working correctly with a big amount specified in mp_maxmoney CVar.

For example: I want to give $100 to the player. When I do it, it sets "-2147483648". You can see the result in the screenshot.

It always occurs when the value of "mp_maxmoney" is higher than "2147483583".

![20190420155538_1](https://user-images.githubusercontent.com/31934972/56458268-c432b080-6384-11e9-91b4-30883055fa59.jpg)

Code:
```
#include <amxmodx>
#include <reapi>

const MONEY = 100;

public plugin_init()
{
	register_clcmd("say /money", "func_Money");
}

public func_Money(id)
{
	client_print_color(id, print_team_default, "^1Current amount of money: ^4%d$", get_member(id, m_iAccount));
	client_print_color(id, print_team_default, "^1Adding:^4 %d$", MONEY);
	rg_add_account(id, MONEY);
	client_print_color(id, print_team_default, "^1New amount of money: ^4%d$", get_member(id, m_iAccount));
	return PLUGIN_HANDLED;
}
```

**Comment by simosalis** [comment]
[quote]
It always occurs when the value of "mp_maxmoney" is higher than "2147483583".
[/quote]
So just don't set an higher value than "2147483583" (maximum positive value allowed for an integer).
Why are you looking for problems?!
Set to something like 1000000000, even less (one less zero), people is far to reach that "naturally"...

**Comment by simosalis** [comment]
"It always occurs when the value of "mp_maxmoney" is higher than "2147483583"."
->
Don't look for problems volontary, so don't set it to an higher value than "2147483583" (maximum positive value of an integer number). By the way, this limit was specified in the readme.md file (the hexadecimal value shouldn't be displayed, because not user-friendly, not nice idea devs..., but you can find the decimal equivalent on the web).
Just set it to 1 000 000 000 (or even one less zero), people is far to be able to reach that amount "naturally"...

**Comment by eat1k** [comment]
https://github.com/s1lentq/ReGameDLL_CS#configuration-cvars
The specified limit is 0x7FFFFFFF = 2147483647. My number is 2147483583. The bug occurs when I write a higher number than this one. I think should be a condition for this.

**Comment by t-nelis** [comment]
@eat1k All cvars are actually represented as single-precision floating point numbers, not 32-bit signed integers. They can represent all integers up to 2²⁴, and after that only multiples of increasing powers of 2.

Say the number you are trying to encode is 2147483584 (2147483583+1) as a single-precision floating point number. The number falls into the 2³⁰+1 to 2³¹ range (log₂2147483584 = 30.999999957) which is relevant for single-precision floating point precision. In this range, only integers that are multiples of 128 can be represented (2^(30-23)).

So, rounding to the nearest multiple of 128 (which is what the program would do while writing the cvar), we get ((2147483584+128÷2)÷128)×128 = 2147483648. Incidentally, log₂2147483648 = 31 exactly (i.e. it is 2³¹). That number is exactly one above the upper bound for 32-bit signed integers or 2³¹-1 (32 bits minus 1 bit for the sign), meaning it cannot be represented in that way (yet this is what the number is converted to when reading the cvar back). Because of 2's complement arithmetics, the variable overflows and the value wraps around to the first negative 32-bit signed integer (−2³¹ or −2147483648, which is what you are seeing in game, albeit truncated because of a lack of screen space).

If I'm right, the actual maximum should be 2147483520 (2³¹−128).

**tl;dr**: This normal and expected.

For simplicity, the documentation could be modified to say that the maximum _precise_ value is 16777216 (2²⁴) instead of 2147483647 (0x7FFFFFFF), although technically higher values can be used provided they are multiples of 2, 4, 8, 16, 32, 64 or 128 depending on how large they are (i.e. in which range they fall).

PS: I haven't tested this so don't take my word for it. You could try with values between 16777216 and 2147483520 which aren't multiples of the powers of 2 mentioned above and hopefully observe that the game displays slightly different values (i.e. values rounded to the nearest multiple of the corresponding power of 2).

**Comment by shelru** [comment]
@s1lentq может быть квар?
Автор видимо не особо понимает что это границы signed int и работает это все как надо.

@eat1k попробуй тоже самое любым c++ интерпретатором.
```cpp
unsigned int a = 0; //Для примера unsigned, можешь использовать signed, но значения другие задать нужно
std::cout<<a;
a-=1;
std::cout<<a;
```

**Comment by eat1k** [comment]
> @s1lentq может быть квар?
> Автор видимо не особо понимает что это границы signed int и работает это все как надо.

@shelru а зачем квар? Я изначально конечно рассчитывал что будет просто лимит до того числа что я написал, но потом увидел что лимит 999к. По сути по началу не понял зачем так надо, ведь есть сервера где используются "миллионы", но потом я понял что это вполне нормально. Лимит на данный момент в игре 999к, именно лимит отображаемого количества денег. Пока он не будет повышен, какой смысл делать это в ReGameDLL/HLDS (там вообще 16к же лимит)?

**Comment by shelru** [comment]
А какой у тебя mp_startmoney был с багом?

> а зачем квар?

Ну у меня на сервере например используется значения больше 10 млн и хранится оно в мембере.
https://github.com/s1lentq/ReGameDLL_CS/commit/430998735c4b112aa9a749bb7d11554e7ddcf823#diff-d02fe19a374f81d1aaee6e4852f59c08R3116

m_iAccount много где фигурирует и его ограничение в 999999 плохо скажется на возможностях разработки.
Мемберов свободных почти не остается, а на модовых серверах они нужны, так как деньги зачастую там скрываются.

Разумнее было бы добавить ограничение на начальный квар.
